crates/store/src/db/models/queries/transactions.rs

```diff
@@ -126,14 +174,210 @@ pub struct TransactionSummaryRowInsert {
+///     )  -- cursor-based pagination
+/// ORDER BY
+///     block_num ASC,
+///     transaction_id ASC
+/// LIMIT
+///     ?6
+/// ```
+/// Notes:
+/// - Uses stable ordering (`block_num`, `transaction_id`) to ensure consistent results across
+///   paginated queries.
+/// - Uses cursor-based pagination.
+/// - The query is executed in chunks of 1000 transactions to prevent loading excessive data and to
+///   stop as soon as the accumulated size approaches the 4MB limit.
+/// - Given the size of note records, 1000 records are guaranteed never to return more than about
+///   60MB of data.
+pub fn select_transactions_records(
+    conn: &mut SqliteConnection,
+    account_ids: &[AccountId],
+    block_range: RangeInclusive<BlockNumber>,
+) -> Result<(BlockNumber, Vec<crate::db::TransactionRecord>), DatabaseError> {
+    const MAX_PAYLOAD_BYTES: i64 = 4 * 1024 * 1024; // 4 MB
+    const NUM_TXS_PER_CHUNK: i64 = 1000; // Read 1000 transactions at a time
+
+    QueryParamAccountIdLimit::check(account_ids.len())?;
+
+    if block_range.is_empty() {
+        return Err(DatabaseError::InvalidBlockRange {
+            from: *block_range.start(),
+            to: *block_range.end(),
+        });
+    }
+
+    let desired_account_ids = serialize_vec(account_ids);
+
+    // Read transactions in chunks to prevent loading excessive data and to stop
+    // as soon as we approach the size limit
+    let mut all_transactions = Vec::new();
+    let mut total_size = 0i64;
+    let mut last_block_num: Option<i64> = None;
+    let mut last_transaction_id: Option<Vec<u8>> = None;
+
+    loop {
+        let mut query =
+            SelectDsl::select(schema::transactions::table, TransactionRecordRaw::as_select())
+                .filter(schema::transactions::block_num.gt(block_range.start().to_raw_sql()))
+                .filter(schema::transactions::block_num.le(block_range.end().to_raw_sql()))
+                .filter(schema::transactions::account_id.eq_any(&desired_account_ids))
+                .into_boxed();
+
+        // Apply cursor-based pagination using the last seen (block_num, transaction_id)
+        if let (Some(last_block), Some(last_tx_id)) = (last_block_num, &last_transaction_id) {
+            query = query.filter(
+                schema::transactions::block_num
+                    .gt(last_block)
+                    .or(schema::transactions::block_num
+                        .eq(last_block)
+                        .and(schema::transactions::transaction_id.gt(last_tx_id))),
+            );
+        }
+
+        let chunk = query
+            .order((
+                schema::transactions::block_num.asc(),
+                schema::transactions::transaction_id.asc(),
+            ))
+            .limit(NUM_TXS_PER_CHUNK)
+            .load::<TransactionRecordRaw>(conn)
+            .map_err(DatabaseError::from)?;
+
+        // If no more data, we're done
+        if chunk.is_empty() {
+            break;
+        }
```

Comment:
nit: I think I'd remove this `if` because it's functionally very similar to the `if` after the `for` loop and it doesn't seem to be too beneficial to return before it since `chunk` would contain no transactions to loop through